pkg/services/x509/x509_validator.go

```diff
@@ -0,0 +1,270 @@
+			break
+		}
+	}
+	if !sigAlgAllowed {
+		return fmt.Errorf("signature algorithm %s is not allowed", x509Token.sigAlg)
+	}
+
+	leafCert, verifiedChains, err := validator.verifyCertChain(x509Token)
+	if err != nil {
+		return err
+	}
+
+	for _, verifiedChain := range verifiedChains {
+		err := validator.checkCertRevocation(verifiedChain)
+		if err != nil {
+			return err
+		}
+	}
+
+	// parse the jwt and verify the jwt signature
+	token, err := jwt.ParseString(x509Token.raw, jwt.WithVerify(x509Token.sigAlg, leafCert.PublicKey))
+	if err != nil {
+		return err
+	}
+
+	// verify the jwt contents such as expiration
+	if err := jwt.Verify(token); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (validator JwtX509Validator) verifyCertChain(x509Token *JwtX509Token) (*x509.Certificate, [][]*x509.Certificate, error) {
+	rootsPool := x509.NewCertPool()
+	for _, cert := range validator.roots {
+		rootsPool.AddCert(cert)
+	}
+	intermediatesPool := x509.NewCertPool()
+	for _, cert := range validator.intermediates {
+		intermediatesPool.AddCert(cert)
+	}
+	verifyOpts := x509.VerifyOptions{
+		Intermediates: intermediatesPool,
+		Roots:         rootsPool,
+		CurrentTime:   time.Now(),
+		KeyUsages:     []x509.ExtKeyUsage{x509.ExtKeyUsageAny},
+	}
+
+	if len(x509Token.chain) == 0 {
+		return nil, nil, fmt.Errorf("token does not have a certificate")
+	}
+
+	for _, cert := range x509Token.chain[1:len(x509Token.chain)] {
+		verifyOpts.Intermediates.AddCert(cert)
+	}
+
+	leafCert := x509Token.chain[0]
+
+	verifiedChains, err := leafCert.Verify(verifyOpts)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to verify certificate chain: %w", err)
+	}
+
+	rootCert := verifiedChains[0][len(verifiedChains[0])-1]
+	if !bytes.Equal(rootCert.RawIssuer, rootCert.RawSubject) || rootCert.CheckSignatureFrom(rootCert) != nil {
+		return nil, nil, fmt.Errorf("certificate is not a root CA")
+	}
+
+	return leafCert, verifiedChains, err
+}
+
+func (validator JwtX509Validator) checkCertRevocation(verifiedChain []*x509.Certificate) error {
+	for i, certToCheck := range verifiedChain[0 : len(verifiedChain)-1] {
+		issuer := verifiedChain[i+1]
+		for _, crlPoint := range certToCheck.CRLDistributionPoints {
+			crl, err := fetchCRL(crlPoint)
+			if err != nil {
+				return fmt.Errorf("could not fetch the crl: %w", err)
+			}
+			if crl.HasExpired(time.Now()) {
+				return fmt.Errorf("crl has been expired")
+			}
+			if err := issuer.CheckCRLSignature(crl); err != nil {
+				return fmt.Errorf("could not check cert agains CRL: %w", err)
+			}
+			revokedCerts := crl.TBSCertList.RevokedCertificates
+			for _, revoked := range revokedCerts {
+				if certToCheck.SerialNumber.Cmp(revoked.SerialNumber) == 0 {
+					return fmt.Errorf("cert with serial '%s' is revoked", certToCheck.SerialNumber.String())
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// fetchCRL fetches and parses a CRL.
+func fetchCRL(url string) (*pkix.CertificateList, error) {
+	resp, err := http.Get(url)
```

Comment:
this looks like hasty written code for calling external services. How long before it times out? (how long does it block the caller)? What if we get 100 concurrent requests for validating a signature with a certificate whose CRL endpoint is timing out after 2 minutes, do we get DDoS-ed? Shouldn't the time-out be configurable?

Reply:
These are good questions, but I think almost a project of their own. I now use the rather naieve http default client and added some documentation to indicate the limitations of the function. https://github.com/nuts-foundation/nuts-auth/issues/136